### ISSUE TITLE
Notebook embed bokeh server not working behind Apache reverse proxy.

### ISSUE THREAD

**Issue by kyang922** [bokeh/bokeh#8475]
#### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)
Bokeh 1.0.1 
Python 3.6.6
Notebook 5.6.0
OS: CentOS Linux 7
browser: Google Chrome 70.0.3538.110

#### Description of expected behavior and the observed behavior
I have a jupyter notebook server running on localhost:8010, and use apache reverse proxy to make it open to internet.  I use the setting example of [JupyterHub](https://bokeh.pydata.org/en/latest/docs/user_guide/notebook.html#jupyterhub), but it get a 400 websocket handshake error and does not work.

I have installed nbserverproxy,  and it looks fine with its hello world example. but when I run my notebook code, chrome console will throw following error.
```
Bokeh: BokehJS plotting callback run at Sat Dec 01 2018 00:40:11 GMT+0900 (Japan Standard Time)
bokeh-1.0.1.min.js:31 [bokeh] setting log level to: 'info'
autoload.js?bokeh-au…3&resources=none:27 Bokeh: all callbacks have finished
bokeh-1.0.1.min.js:31 WebSocket connection to 'wss://jupyter.xxx.com/proxy/38633/ws?bokeh-protocol-version=1.0&bokeh-session-id=xxx' failed: Error during WebSocket handshake: Unexpected response code: 400
bokeh-1.0.1.min.js:31 [bokeh] Failed to connect to Bokeh server Error: Could not open websocket
bokeh-1.0.1.min.js:31 [bokeh] Lost websocket 0 connection, 1006 ()
kernel.js:1007 Couldn't process kernel message SyntaxError: Unexpected token W in JSON at position 0
    at JSON.parse (<anonymous>)
    at html_table (main.js?v=20181130181509:130)
    at code_exec_callback (main.js?v=20181130181509:168)
    at Kernel._handle_output_message (kernel.js:1196)
    at i (jquery.min.js:2)
    at Kernel._handle_iopub_message (kernel.js:1223)
    at Kernel._finish_ws_message (kernel.js:1015)
    at kernel.js:1006
bokeh-1.0.1.min.js:31 [bokeh] Websocket connection 0 disconnected, will not attempt to reconnect
```
and jupyter notebook terminal output following message
```
[I 00:40:08.981 NotebookApp] Kernel restarted: eb8cfb67-55c5-4942-b586-86d20801eed9
[I 00:40:09.667 NotebookApp] Adapting to protocol v5.1 for kernel eb8cfb67-55c5-4942-b586-86d20801eed9
[I 00:40:09.668 NotebookApp] Restoring connection for eb8cfb67-55c5-4942-b586-86d20801eed9:d4da500b5dd446daaae5eeffb167d0ce
[I 00:40:09.668 NotebookApp] Replaying 3 buffered messages
WARNING:tornado.access:400 GET /ws?bokeh-protocol-version=1.0&bokeh-session-id=xxx (::1) 0.50ms
[W 00:40:11.276 NotebookApp] 400 GET /proxy/38633/ws?bokeh-protocol-version=1.0&bokeh-session-id=xxx (::1) 4.79ms referer=None

```
#### Complete, minimal, self-contained example code that reproduces the issue

------ notebook ------
 ```
import yaml

from bokeh.layouts import column
from bokeh.models import ColumnDataSource, Slider
from bokeh.plotting import figure
from bokeh.themes import Theme
from bokeh.io import show, output_notebook

from bokeh.sampledata.sea_surface_temperature import sea_surface_temperature

output_notebook()

def modify_doc(doc):
    df = sea_surface_temperature.copy()
    source = ColumnDataSource(data=df)

    plot = figure(x_axis_type='datetime', y_range=(0, 25),
                  y_axis_label='Temperature (Celsius)',
                  title="Sea Surface Temperature at 43.18, -70.43")
    plot.line('time', 'temperature', source=source)

    def callback(attr, old, new):
        if new == 0:
            data = df
        else:
            data = df.rolling('{0}D'.format(new)).mean()
        source.data = ColumnDataSource(data=data).data

    slider = Slider(start=0, end=30, value=0, step=1, title="Smoothing by N Days")
    slider.on_change('value', callback)

    doc.add_root(column(slider, plot))

    doc.theme = Theme(json=yaml.load("""
        attrs:
            Figure:
                background_fill_color: "#DDDDDD"
                outline_line_color: white
                toolbar_location: above
                height: 500
                width: 800
            Grid:
                grid_line_dash: [6, 4]
                grid_line_color: white
    """))

import os
import urllib

def remote_jupyter_proxy_url(port):
    """
    Callable to configure Bokeh's show method when a proxy must be
    configured.

    If port is None we're asking about the URL
    for the origin header.
    """
    base_url = 'https://jupyter.xxx.com'
    host = urllib.parse.urlparse(base_url).netloc

    # If port is None we're asking for the URL origin
    # so return the public hostname.
    if port is None:
        return host

    service_url_path = ''
    proxy_url_path = 'proxy/%d' % port

    user_url = urllib.parse.urljoin(base_url, service_url_path)
    full_url = urllib.parse.urljoin(user_url, proxy_url_path)
    return full_url

show(modify_doc, notebook_url=remote_jupyter_proxy_url)
```

------ Apache setting ------
```
<VirtualHost _default_:443>
    ServerName jupyter.xxx.com
    ServerAdmin webmaster@xxx.com

    ProxyRequests Off 
    ProxyPreserveHost On

    <Location />
        ProxyPass http://localhost:8010/
        ProxyPassReverse http://localhost:8010/
    </Location>

    <Location /terminals/websocket/>
        ProxyPass ws://localhost:8010/terminals/websocket/
        ProxyPassReverse ws://localhost:8010/terminals/websocket/
    </Location>

    <LocationMatch /(api/kernels/[^/]+/channels.*)>
        ProxyPass ws://localhost:8010/$1
        ProxyPassReverse ws://localhost:8010/$1
    </LocationMatch>

    ErrorLog logs/jupyter-notebook-error.log
    LogLevel warn
    CustomLog logs/jupyter-notebook-access.log combined
    
    SSLEngine on
    SSLProtocol all -SSLv3 -TLSv1
    SSLHonorCipherOrder on
    SSLCipherSuite HIGH:MEDIUM:!MD5:!RC4

    SSLCertificateFile /etc/letsencrypt/live/xxx.com/fullchain.pem
    SSLCertificateKeyFile /etc/letsencrypt/live/xxx.com/privkey.pem
    
    BrowserMatch "MSIE [2-5]" \
        nokeepalive ssl-unclean-shutdown \
        downgrade-1.0 force-response-1.0
</VirtualHost>
```

**Comment by bryevdv** [comment]
I'm afraid I don't have the expertise to really help, I have never used Apache as a reverse proxy. websockets. Some light googling indicates the error message indicates a [configuration issue with he proxy](https://github.com/socketio/socket.io/issues/1942), so I am going to close this and refer you to the  [public mailing list](https://groups.google.com/a/continuum.io/forum/#!forum/bokeh and  [Stack Overflow](https://stackoverflow.com/) for support. There is a larger audience in both places, and thus a higher probability that someone with Apache experience will be present to help. If concrete guidance for Apache arises we would be happy to add them to our documentation.

**Comment by kyang922** [comment]
I have got this problem solved. it was caused by inappropriate Apache settings. When bokeh use `wss://jupyter.xxx.com/proxy/38633/ws?bokeh-protocol-version=1.0&bokeh-session-id=xxx` to make a websocket connection to the behind server,  this URL will first be parsed by Apache, then parsed by nbserverproxy.

when  this URL is being parsed by Apache, it will match
```
<Location />
    ProxyPass http://localhost:8010/
    ProxyPassReverse http://localhost:8010/
</Location>
```
so apache will make a proxy connection to nbserverproxy using http protocol, then nbserverproxy will also make a proxy connection with backend bokeh server using same protocol.


    browser --- (wss) --- Apache --- (http) --- nbserverproxy --- (http) -X- bokeh server


since backend server is expected to receive a websocket connection, so it output a 400 error.

I simply add a `Location` to get  this problem corrected.
```
<LocationMatch "^/(proxy/[0-9]+/ws.*)">
    ProxyPass "ws://localhost:8010/$1"
    ProxyPassReverse "ws://localhost:8010/$1"
</LocationMatch>
```
this will make connection as following and it works fine.

    browser --- (wss) --- Apache --- (ws) --- nbserverproxy --- (ws) --- bokeh server
